tests/test_users_utils.py

```diff
@@ -0,0 +1,107 @@
+import unittest
+from unittest.mock import patch, MagicMock
+from flaskblog import create_app, db, bcrypt
+from flaskblog.models import User
+from flaskblog.users.utils import save_picture, remove_old_picture, send_reset_email
+from PIL import Image
+from flask import url_for
+
+
+class TestBase(unittest.TestCase):
+    def setUp(self):
+        self.app = create_app()
+        self.app.config["TESTING"] = True
+        self.app.config["WTF_CSRF_ENABLED"] = False
+        self.app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///:memory:"
+        self.app.config["SERVER_NAME"] = "localhost.localdomain"
+        self.client = self.app.test_client()
+        with self.app.app_context():
+            db.create_all()
+            hashed_password = bcrypt.generate_password_hash("password")
+            user = User(
+                username="testuser",
+                email="test@example.com",
+                password=hashed_password,
+                email_verified=True,
+                first_name="Test_first_name",
+                last_name="Test_last_name",
+                role="Follower",
+            )
+            db.session.add(user)
+            db.session.commit()
+
+    def tearDown(self):
+        with self.app.app_context():
+            db.session.remove()
+            db.drop_all()
+
+
+class TestUtils(TestBase):
+
+    def test_save_picture(self):
```

Comment:
These are "happy path tests". 
"Happy Path" tests exercise the function/method as its expected to work.
Your code currently doesn't handle errors and, therefore, your tests can't handle what happens if one of the methods throws an error.

For example, what if the `Image.Open()` method throws an error?
Do you want that error to be handled the same as if `Image.Save()` failed? Maybe, maybe not.
That is the purpose of unit tests. To makes sure the happy path works, but also all the errors are handled properly.

Another purpose of unit tests is to test all the possible combinations of function/method inputs.
What if `none` was passed to `save_picture()`? Is this error caught and handled?

And, finally, logging. You want to make sure handled errors are logged so you can debug them.
Most of the time the default error message isn't helpful at all. No insight into the nature of the failure and you have to guess where the problem _might_ be.

You built a lot very fast (the downside of tutorials).
And when you built it you weren't considering tests.